Automattic/simplenote-android
# Add Checklists Feature

Adds a UI for "checkable" checklists in the note editor. Behind the scenes, we parse markdown checklist syntax (e.g. "- [ ]") and convert them to ClickableSpans. When the note saves, it parses the spannable string and converts it back to the markdown syntax. The app will also display the checklist icons in the notes list.

<img width="306" alt="screen shot 2018-12-21 at 12 08 04 pm" src="https://user-images.githubusercontent.com/789137/50361627-179d0b00-0519-11e9-9cb5-e5bde743d93d.png">

Tapping on a checklist item will add or remove the `x` to the note content when it saves.

There were some unfortunate rabbit holes when making this :) There's some new subclasses:

* `CheckableSpan`: Includes a boolean to track whether the checkbox is ticked or not and to display the appropriate drawable.
* `SimplenoteMovementMethod`: Allows the `CheckableSpan`s to be tapped, but doesn't fire the onClick() listener for any other spans. This is so the regular link detector still works as expected (placing the cursor inside a web url for example should show the context action buttons in the toolbar)
* `CenteredImageSpan`: Centers the drawable with the text, so the checkbox aligns nicely.
* There's also a lot of work to figure out how to highlight the search results properly. We need to adjust the index of the matches so that they still highlight correctly after the note content has been replaced with the `CheckableSpan`s.

**To Test**
* Open a note, and create a checklist! Pressing `return` at the end of a checklist item should automatically add a new one on the next line.
* Open the note on app.simplenote.com, it should have the correct markdown syntax displayed.
* Update the note on the web browser, the app should update and display any modified checklists appropriately.
* Test things that modify the note content, like the history browser and ensure they work as expected.
* Run the new tests!

Refs: #528 #605
